### ISSUE TITLE
bar_label clipping

### ISSUE THREAD

**Issue by anntzer** [matplotlib/matplotlib#19429]
### Bug report

**Bug summary**

bar_label currently uses the default annotation() clipping behavior, which is to draw the text unclipped as long as the annotated point is within the axes.  I'm not sure this is the best behavior for this use case?

**Code for reproduction**

```python
from pylab import *

ax = gca()
rs = ax.bar(np.arange(50), np.arange(50))
ax.bar_label(rs)
ax.set(ylim=(0, 40))
show()
```

**Actual outcome**

![test](https://user-images.githubusercontent.com/1322974/106592644-1344d800-6550-11eb-8825-828e34b718ea.png)

Note how the 39 and 40 labels are still drawn (effectively, because the bar tops are within axes limits), but not the subsequent ones.

**Expected outcome**

I'm not sure what the best behavior would be, but it should likely be one of the following:
- perform "normal" clipping (so 39 is half-clipped, 40 is fully clipped out)
- don't draw any labels that would be clipped (39 is hidden)
- don't clip at all (all labels are visible)

bar_label is new in 3.4 so it *may* be worth deciding on this before the 3.4 release.

**Matplotlib version**

  * Operating system: linux
  * Matplotlib version (`import matplotlib; print(matplotlib.__version__)`): head
  * Matplotlib backend (`print(matplotlib.get_backend())`): agg
  * Python version: 39
  * Jupyter version (if applicable): 
  * Other libraries:

**Comment by jklymak** [comment]
Is there a reason this is release critical?

**Comment by anntzer** [comment]
bar_label is new in 3.4, so the question is just whether we'd have to go through a deprecation cycle if we decide to change the behavior after 3.4 (OTOH we could also deem the change (if we decide for one) to be a bugfix).

**Comment by jklymak** [comment]
#15602 ping @immaxchen  did you have any input on this?

**Comment by timhoffm** [comment]
IMHO 

> - perform "normal" clipping (so 39 is half-clipped, 40 is fully clipped out)

is the right way to go. By default nothing should leak out of the Axes.

But labeled bars that go out of limits is rather an edge case. Usually, you would like to see all your labels. So I'm also ok to fix this in 3.4.x+ and have it as a minor bug until then. This should not hold up the 3.4 release.

**Comment by immaxchen** [comment]
yes, `bar_label` uses the default `annotate` clipping behavior because it use `annotate` under the hood
this default behavior is fairly reasonable to me since I would expect intact labels in almost all my use cases
however, you could always call `bar_label` with `clip_on=True` to get half-clipped labels, since `kwargs` will be passed through
I'd not regard this as a bug though, it's more like a matter of choice (and maybe change should be made on `annotate` not here?)

**Comment by tacaswell** [comment]
> perform "normal" clipping (so 39 is half-clipped, 40 is fully clipped out)

Do we have the ability to do that with text?

**Comment by anntzer** [comment]
yes (try `text(0, 0, "foo", ha="center", va="center", clip_on=True, size=42)`)
